### PR TITLE
V4 api

### PR DESCRIPTION
**Pull request by fdelbos**
update to v4 api

### CODE REVIEW

File: teamweek/teamweek.go [hunk 1]
```diff
@@ -1,191 +1,173 @@
...
+		EndTime          string   `json:"end_time,omitempty"`
+		Color            string   `json:"color,omitempty"`
+		ColorID          int64    `json:"color_id,omitempty"`
+		EstimatedMinutes int64    `json:"estimated_minutes,omitempty"`
+		Done             bool     `json:"done,omitempty"`
+		Pined            bool     `json:"pinned,omitempty"`
+		UserID           int64    `json:"user_id,omitempty"`
+		FolderID         int64    `json:"folder_id,omitempty"`
+		Position         int64    `json:"position,omitempty"`
+		Weight           int64    `json:"weight,omitempty"`
```
Comment (shedokan) [inline teamweek/teamweek.go]: `weight` is returned as a float, `"weight":0.0`:
```
panic: json: cannot unmarshal number 0.0 into Go struct field Task.weight of type int64
```

File: teamweek/teamweek.go [hunk 2]
```diff
@@ -1,191 +1,173 @@
...
-		UpdatedAt   string        `json:"updated_at,omitempty"`
+	Task struct {
+		ID               int64    `json:"id,omitempty"`
+		Name             string   `json:"name,omitempty"`
+		Notes            string   `json:"notes,omitempty"`
+		StartDate        string   `json:"start_date,omitempty"`
+		EndDate          string   `json:"end_date,omitempty"`
+		StartTime        string   `json:"start_time,omitempty"`
+		EndTime          string   `json:"end_time,omitempty"`
+		Color            string   `json:"color,omitempty"`
```
Comment (shedokan) [inline teamweek/teamweek.go]: Color is returned as an integer `"color_id":29,"color":29`:
```
panic: json: cannot unmarshal number into Go struct field Task.color of type string
```

File: teamweek/teamweek.go [hunk 3]
```diff
@@ -1,191 +1,173 @@
...
-		Color          string   `json:"color,omitempty"`
-		EstimatedHours float64  `json:"estimated_hours,omitempty"`
-		Pinned         bool     `json:"pinned,omitempty"`
-		Done           bool     `json:"done,omitempty"`
-		UserID         int64    `json:"user_id,omitempty"`
-		Project        *Project `json:"project,omitempty"`
+		ID        int64  `json:"id,omitempty"`
+		Name      string `json:"name,omitempty"`
+		Color     string `json:"color,omitempty"`
+		ColorID   string `json:"color_id,omitempty"`
```
Comment (shedokan) [inline teamweek/teamweek.go]: Should be `int64`

File: teamweek/teamweek.go [hunk 4]
```diff
@@ -1,191 +1,173 @@
...
-		EndTime        string   `json:"end_time,omitempty"`
-		Color          string   `json:"color,omitempty"`
-		EstimatedHours float64  `json:"estimated_hours,omitempty"`
-		Pinned         bool     `json:"pinned,omitempty"`
-		Done           bool     `json:"done,omitempty"`
-		UserID         int64    `json:"user_id,omitempty"`
-		Project        *Project `json:"project,omitempty"`
+		ID        int64  `json:"id,omitempty"`
+		Name      string `json:"name,omitempty"`
+		Color     string `json:"color,omitempty"`
```
Comment (shedokan) [inline teamweek/teamweek.go]: Color is returned as an integer `"color_id":29,"color":29`:
```
panic: json: cannot unmarshal number into Go struct field Project.color of type string
```

File: teamweek/teamweek.go [hunk 5]
```diff
@@ -1,191 +1,173 @@
...
-		Initials   string `json:"initials,omitempty"`
-		PictureUrl string `json:"picture_url,omitempty"`
-		HasPicture bool   `json:"has_picture,omitempty"`
-		Color      string `json:"color,omitempty"`
-		Weight     int64  `json:"weight,omitempty"`
-		Dummy      bool   `json:"dummy,omitempty"`
+	Milestone struct {
+		ID        int64  `json:"id,omitempty"`
+		Name      string `json:"name,omitempty"`
+		Color     string `json:"color,omitempty"`
```
Comment (shedokan) [inline teamweek/teamweek.go]: Color is returned as an integer `"color_id":29,"color":29`:
```
panic: json: cannot unmarshal number into Go struct field Milestone.color of type string
```

File: teamweek/teamweek.go [hunk 6]
```diff
@@ -1,191 +1,173 @@
...
+		Workspaces []Workspace `json:"workspaces,omitempty"`
+		ProjectIDs []int64     `json:"project_ids,omitempty`
+		CreatedAt  string      `json:"created_at,omitempty"`
+		UpdatedAt  string      `json:"updated_at,omitempty"`
+	}
+
+	Member struct {
+		Role              string `json:"role,omitempty"`
+		Active            bool   `json:"active,omitempty"`
+		Color             string `json:"color,omitempty"`
```
Comment (shedokan) [inline teamweek/teamweek.go]: `color` is not returned for members

File: teamweek/teamweek.go [hunk 7]
```diff
@@ -1,191 +1,173 @@
...
+		ID         int64       `json:"id,omitempty"`
+		Name       string      `json:"name,omitempty"`
+		Email      string      `json:"email,omitempty"`
+		Color      string      `json:"color,omitempty"`
+		ColorID    int64       `json:"color_id,omitempty"`
+		Initials   string      `json:"initials,omitempty"`
+		PictureUrl string      `json:"picture_url,omitempty"`
+		HasPicture bool        `json:"has_picture,omitempty"`
+		Workspaces []Workspace `json:"workspaces,omitempty"`
+		ProjectIDs []int64     `json:"project_ids,omitempty`
```
Comment (shedokan) [inline teamweek/teamweek.go]: Missing `"` at the end

File: teamweek/client.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+func (c *Client) put(url string, dest, body interface{}) error {
+	return c.request(url, http.MethodPut, dest, body)
+}
+
+func (c *Client) remove(url string) error {
+	return c.request(url, http.MethodDelete, nil, nil)
+}
+
+func (c *Client) request(urlStr, method string, dest, body interface{}) error {
```
Comment (shedokan) [inline teamweek/client.go]: Nit picking: As `dest` is optional it seems, maybe have it last?

File: teamweek/client.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+		return ErrForbidden
+
+	case http.StatusNotFound: // 404
+		return ErrNotFound
+
+	case http.StatusInternalServerError: // 500
+		return ErrInternal
+
+	default:
+		if resp.StatusCode >= 400 { // for every other error
```
Comment (shedokan) [inline teamweek/client.go:92]: Why is this last one needs to be inside `default:`?

File: teamweek/client.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	client    *http.Client
+	BaseURL   *url.URL
+	UserAgent string
+}
+
+func (c *Client) get(url string, dest interface{}) error {
+	return c.request(url, http.MethodGet, dest, nil)
+}
+
+func (c *Client) post(url string, dest, body interface{}) error {
```
Comment (shedokan) [inline teamweek/client.go]: What's the point of these **private** methods if they are not used within the library?

**Review by shedokan — CHANGES_REQUESTED** [review]
Some of what I've written are TW bugs, either in the docs or API, for now we can comment those fields out and please recheck them and notify TW to fix those bugs.

**Comment by fdelbos** [comment]
I ll remove the color fields and other ambiguous fields as they are not needed

File: teamweek/client.go (hunk 10 above)
Comment (fdelbos) [reply teamweek/client.go]: I was hoping i could write a real lib but, no time for that...

File: teamweek/client.go (hunk 8 above)
Comment (fdelbos) [reply teamweek/client.go]: body is optional for get and delete (i was hoping to do post and put...) but the dest is optionnal only for delete so i think we should keep it like that.

File: teamweek/teamweek.go (hunk 7 above)
Comment (fdelbos) [reply teamweek/teamweek.go]: good catch!

File: teamweek/client.go (hunk 10 above)
Comment (shedokan) [reply teamweek/client.go]: You need a usecase for that, otherwise it'll get stale.

Comment (fdelbos) [reply teamweek/client.go]: I ll remove all that and hopefully some day could finish it.

**Review by shedokan — APPROVED** [review]
One refactoring comment wasn't dealt with, but overall LGTM

**Pull request merged**